src/getting-started.md

```diff
@@ -0,0 +1,406 @@
+caches.
+
+### Cloning
+
+You can just do a normal git clone:
+
+```shell
+git clone https://github.com/rust-lang/rust.git
+```
+
+You don't need to clone the submodules at this time. But if you want to, you
+can do the following:
+
+```sh
+# first time
+git submodule update --init --recursive
+
+# subsequent times (to pull new commits)
+git submodule update
+```
+
+**Pro tip**: if you contribute often, you may want to look at the git worktrees
+tip in [this chapter][suggested].
+
+### Configuring the Compiler
+
+The compiler has a configuration file which contains a ton of settings. We will
+provide some recommendations here that should work for most, but [check out
+this chapter for more info][config].
+
+[config]: ./building/how-to-build-and-run.html#create-a-configtoml
+
+In the top level of the repo:
+
+```shell
+cp config.toml.example config.toml
+```
+
+Then, edit `config.toml`. You will need to search for, uncomment, and update
+the following settings:
+
+- `debug = true`: enables debug symbols and `debug!` logging, takes a bit longer to compile.
+- `incremental = true`: enables incremental compilation of the compiler itself.
+  This is turned off by default because it's technically unsound. Sometimes
+  this will cause weird crashes, but it can really speed things up.
+- `llvm-config`: enables building with system LLVM. [See this chapter][sysllvm]
+  for more info. This avoids building LLVM, which can take a while.
+
+[sysllvm]: ./building/suggested.html#building-with-system-llvm
+
+### `./x.py` Intro
+
+`rustc` is a bootstrapping compiler because it is written in Rust. Where do you
+get the original compiler from? We use the current beta compiler
+to build the compiler. Then, we use that compiler to build itself. Thus,
+`rustc` has a 2-stage build.
+
+We have a special tool `./x.py` that drives this process. It is used for
+compiling the compiler, the standard libraries, and `rustdoc`. It is also used
+for driving CI and building the final release artifacts.
+
+### Building and Testing `rustc`
+
+For most contributions, you only need to build stage 1, which saves a lot of time.
+After updating `config.toml`, as mentioned above, you can use `./x.py`:
+
+```shell
+# Build the compiler (stage 1)
+./x.py build --stage 1
+
+# Subsequent builds
+./x.py build --stage 1 --keep-stage 1
+```
+
+This will take a while, especially the first time. Be wary of accidentally
+touching or formatting the compiler, as `./x.py` will try to recompile it.
+
+**NOTE**: The `--keep-stage 1` will _assume_ that the stage 0 standard library
+does not need to be rebuilt, which is usually true, which will save some time.
+However, if you are changing certain parts of the compiler, this may lead to
+weird errors. Feel free to ask on [zulip][z] if you are running into issues.
+
+To run the compiler's UI test suite (the bulk of the test suite):
+
+```
+# UI tests
+./x.py test --stage 1 src/test/ui [--keep-stage 1]
```

Comment:
I suggest separating this into two lines like above, because it might be a little bit confusing otherwise.

```suggestion
# First build
./x.py test --stage 1 src/test/ui

# Subsequent builds
./x.py test --stage 1 src/test/ui --keep-stage 1
```

Reply:
The `--keep-stage 1` option confused me when I was getting started - I wasn't sure whether it was the compiler that was being kept, or some build of the test suite, or something else. I don't know if that's something other people run into, but I wanted to mention it just in case.

Reply:
Yeah, the naming is not ideal... also the numbering of the stages...